Adds the extended keycode where appropriate and updates to the latest IronRDP master commit hash.

ibeckermayer opened the pull request:
Both of these changes are necessary in order to get the windows key to work again.

Corresponds with https://github.com/Devolutions/IronRDP/pull/364

changelog: Allow users to send the Windows key for desktop sessions.

Commit:
Adds the extended keycode where appropriate and updates to the latest IronRDP master commit hash.

ibeckermayer commented on lib/srv/desktop/rdp/rdpclient/src/client.rs:
```diff
@@ -447,6 +464,11 @@ impl Client {
         if !key.down {
             flags = KeyboardFlags::RELEASE;
         }
+        let extended = key.code & 0xE000 == 0xE000;
+        if extended {
+            flags |= KeyboardFlags::EXTENDED;
+        }
```
This is where we actually set the `EXTENDED` flag; the rest of the changes are to interface with the updated IronRDP apis.

zmb3 reviewed (approved):
Seems fine to me. Let's get the disconnect reason plumbed through to the UI soon though. That was a really nice enhancement when we added it for rdp-rs

Commit:
cleaning up disconnect result

Commit:
Pipes disconnect messages through to the client.

This required a change to the state management for DesktopSession. State
is still quite complex to get right here, but this is a step in the right
direction.

Commit:
Updates desktopPlaybackHandle to the new authenticated websocket paradigm as it was forgotten in https://github.com/gravitational/teleport/pull/37520

ibeckermayer commented:
@zmb3 @probakowski Please take another look.

I ended up needing to include some pretty substantive changes in order to pipe messages through to the client itself. I also realize we forgot to swap in the desktopplayback endpoint in https://github.com/gravitational/teleport/pull/37962 new authenticated websocket paradigm, so those changes are included now as well.

Commit:
prettier formatting

Commit:
adding clarifying comments

probakowski commented:
Are changes around playback ws the same here as in https://github.com/gravitational/teleport/pull/37981? If so would it make sense to based this PR on top of that one?

Commit:
Merge branch 'master' into isaiah/update-ironrdp-latest

zmb3 commented on web/packages/teleport/src/DesktopSession/DesktopSession.tsx:
```diff
@@ -58,92 +51,97 @@ declare global {
   }
 }
 
+const invalidStateMessage =
```
I don't know how I feel about having the product say "something went wrong, please file a bug report." Can we just display a more generic error and log whatever information we can?

ibeckermayer replied:
Changed to `const invalidStateMessage = 'internal application error';` in 6cfb564322a1deedff79e1451b2d5982d8f92bbe

zmb3 commented on web/packages/teleport/src/DesktopSession/DesktopSession.tsx:
```diff
@@ -58,92 +51,97 @@ declare global {
+          } else if (tdpConnection.status === '') {
+            message =
+              tdpConnection.statusText || 'TDP connection ended gracefully';
+          } else if (wsConnection.status === 'closed') {
```
The number of nested if/else blocks in here is 🤯

Any way we can simplify this?

ibeckermayer replied:
I did another minor overhaul to make the state machine more comprehensible: 6cfb564322a1deedff79e1451b2d5982d8f92bbe

zmb3 commented on web/packages/teleport/src/Player/DesktopPlayer.tsx:
```diff
@@ -175,6 +177,14 @@ const useDesktopPlayer = ({ clusterId, sid }) => {
+      setPlayerStatus(StatusEnum.COMPLETE);
+      setStatusText(info);
+    },
+    [setStatusText]
```
Why do we include `setStatuxText` but not `setPlayerStatus`?

In any case, we can probably omit both. The React hooks docs say:
> React guarante
es that setState function identity is stable and won’t change on re-renders. This is why it’s safe to omit from the useEffect or useCallback dependency list.

ibeckermayer replied:
c4c4aa2c6e2cad87db660351b95cc66391bdaf46

zmb3 commented on web/packages/teleport/src/lib/tdp/client.ts:
```diff
@@ -134,7 +136,11 @@ export default class Client extends EventEmitterWebAuthnSender {
+    this.socket.onclose = ev => {
+      let message = 'session disconnected';
+      if (ev.code !== WebsocketCloseCode.NORMAL) {
+        message = `connection closed with websocket error code: ${ev.code}`;
```
Users aren't going to have any idea what various websocket error codes mean. This seems like something that could be logged but doesn't need to be user-facing.

ibeckermayer replied:
Updated in 6cfb564322a1deedff79e1451b2d5982d8f92bbe

Commit:
Further refactoring to make DesktopSession's state machine more comprehensible

Commit:
Remove unnecessary items from dependency lists

zmb3 commented on lib/srv/desktop/rdp/rdpclient/client.go:
```diff
@@ -305,14 +310,37 @@ func (c *Client) startRustRDP(ctx context.Context) error {
+			err = trace.Errorf("RDP client exited with an unknown error")
+		}
+
+		c.sendTDPNotification(err.Error(), tdp.SeverityError)
```
This is okay for now, but Rust errors are generally not suitable for display in the UI. Let's file an action item to clean this up.

ibeckermayer replied:
https://github.com/gravitational/teleport/issues/38373

zmb3 commented on lib/srv/desktop/rdp/rdpclient/client.go:
```diff
@@ -305,14 +310,37 @@ func (c *Client) startRustRDP(ctx context.Context) error {
+	}
+
+	if message != "" {
+		message = fmt.Sprintf("RDP client exited gracefully with message: %v", message)
```
This is better than what we had, but still not super user friendly:
<img width="565" alt="image" src="https://github.com/gravitational/teleport/assets/7527103/6083a7fe-9536-4eb4-b2e7-7acdb8eafb43">

With rdp-rs we had the lookup table with human readable messages for each disconnect code: 

https://github.com/gravitational/rdp-rs/blob/edfb5330a11d11eaf36d65e4300555368b4c6b02/src/core/global.rs#L309-L339

Doesn't need to be addressed in this PR but I think we should get it done soon.

ibeckermayer replied:
https://github.com/gravitational/teleport/issues/38373

Commit:
Merge branch 'master' into isaiah/update-ironrdp-latest

ibeckermayer commented:
> Are changes around playback ws the same here as in #37981? If so would it make sense to based this PR on top of that one?

That may have been the case when you asked the question, but everything should be up to date now.
